8.0 'stock' (dev)

legalsylvain opened the pull request:
Port 'stock' modification realized by OneStein BV. (branch onesteinbv:8.0-os)

legalsylvain committed:
[ADD] [8.0] 'stock' migration;

red15 commented:
Getting error with `openupgrade_legacy_8_0_procurement_id` not existing in model `stock.warehouse.orderpoint`.

I think you might need to add the 'procurement_id' column rename:

```
--- a/addons/stock/migrations/8.0.1.1/pre-migration.py
+++ b/addons/stock/migrations/8.0.1.1/pre-migration.py
@@ -64,6 +64,7 @@ column_renames = {
     ],
     'stock_warehouse_orderpoint': [
         ('product_uom', None),
+        ('procurement_id', None),
     ],
 }
```

red15 commented:
Also missing `product.template / procure_method`

hbrunn commented:
@red15 this depends on #121, there both renames are done

red15 commented:
@hbrunn well spotted, I was trying this branch alone to avoid mixing responsibilities. It's a good thing this huge work they have done is split up. Unfortunately this means you can get missing things like this.

legalsylvain commented:
@red15 :
Procurement module depends on 'stock' module. So there is no problem. When the 'stock' migration will be reviewed. (and it has to be reviewed first), so the migration of the procurement module will work.

Regards.

legalsylvain commented:
Oh wait, I spoke too soon.

In this case, we maybe have to move part of the 'procurement' (#121) migration code into this PR.

A migration of stock without procurement module installed has to work.

hbrunn commented:
@legalsylvain it's the other way around, right? stock depends on procurement, so in my opinion, that's fine

legalsylvain commented:
It depends ! We are both right. I checked V7 and you checked V8.
- V7: 'procurement' depends on 'stock':
  https://github.com/OCA/OCB/blob/7.0/addons/procurement/__openerp__.py#L29
- V8: 'stock' depends on 'procurement':
  https://github.com/OCA/OCB/blob/8.0/addons/stock/__openerp__.py#L55

-> basically, if procurement is not installed on V7, We have to install it ... and maybe before the migration. Could we imagine that OpenUpgrade block migration if procurement is not installed (and stock is). It will be the most simple. I'm not sure...

@StefanRijnhart, @hbrunn, @all, what do you think about that point ?

Regards.

StefanRijnhart commented:
Well spotted! Procurement should be installed automatically on databases with stock installed in 7.0 without any work from our side. But the stock migration script should then not assume the availabilty of any legacy tables or columns from the procurement module (haven't checked if it does).

hbrunn commented:
it does, that's the problem @legalsylvain raised earlier. Here my understanding:

| stock | procurement | problem? |
| --- | --- | --- |
| yes | yes | no |
| yes | no | yes |
| no | yes | no (shouldn't happen) |
| no | no | no (trivial) |

So to handle the second case, all references to legacy procurement columns should first check if they exist in the fist place.

legalsylvain commented:
I'm lost.
Why " But the stock migration script should then not assume the availabilty of any legacy tables or columns from the procurement module (haven't checked if it does).", if we block migration of database with 'stock' installed and 'procurement' uninstalled.
In that case, we can assume that second case will not occure.
And so, use procurement column in stock migration...

(I'm not sure that it is the good solution, but for me the most simple prima facie.)

StefanRijnhart commented:
The stock migration scripts should simply check if the procurement legacy table and/or columns that it wants to use exist, and apply a strategy for when they don't.

legalsylvain commented:
Agree, it's the most modular solution.

hbrunn committed:
[FIX] check for existence of legacy columns from procurement

hbrunn commented:
my latest commit should do the trick

christophlsa commented:
I have some errors according to changing colums which have null values. Am I right that the workflow of migration is the following?:
1. pre-migration
2. normal odoo module update
3. post-migration

When this is right then I think some methods from post-migration have to be moved to pre-migration. Or have I only rerun the update script and it's ok (is this a good workflow?)? I looked in the tables in database and these null values are filled after post-migration (too late for the module update).

Here are my errors:

```
2014-12-12 09:32:31,017 23335 ERROR testdb openerp.sql_db: bad query: ALTER TABLE "stock_picking" ALTER COLUMN "picking_type_id" SET NOT NULL
Traceback (most recent call last):
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "picking_type_id" contains null values

2014-12-12 09:32:31,018 23335 WARNING testdb openerp.models: WARNING: unable to set column picking_type_id of table stock_picking not null !
Try to re-run: openerp-server --update=module
If it doesn't work, update records and execute manually:
ALTER TABLE stock_picking ALTER COLUMN picking_type_id SET NOT NULL
Traceback (most recent call last):
  File "/var/tmp/openupgrade/8.0/server/openerp/models.py", line 2702, in _auto_init
    cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, k))
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 158, in wrapper
    return f(self, *args, **kwargs)
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "picking_type_id" contains null values

2014-12-12 09:32:36,946 23335 ERROR testdb openerp.sql_db: bad query: ALTER TABLE "stock_warehouse" ALTER COLUMN "code" SET NOT NULL
Traceback (most recent call last):
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "code" contains null values

2014-12-12 09:32:36,947 23335 WARNING testdb openerp.models: WARNING: unable to set column code of table stock_warehouse not null !
Try to re-run: openerp-server --update=module
If it doesn't work, update records and execute manually:
ALTER TABLE stock_warehouse ALTER COLUMN code SET NOT NULL
Traceback (most recent call last):
  File "/var/tmp/openupgrade/8.0/server/openerp/models.py", line 2702, in _auto_init
    cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, k))
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 158, in wrapper
    return f(self, *args, **kwargs)
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "code" contains null values

2014-12-12 09:32:36,991 23335 ERROR testdb openerp.sql_db: bad query: ALTER TABLE "stock_warehouse" ALTER COLUMN "view_location_id" SET NOT NULL
Traceback (most recent call last):
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "view_location_id" contains null values

2014-12-12 09:32:36,991 23335 WARNING testdb openerp.models: WARNING: unable to set column view_location_id of table stock_warehouse not null !
Try to re-run: openerp-server --update=module
If it doesn't work, update records and execute manually:
ALTER TABLE stock_warehouse ALTER COLUMN view_location_id SET NOT NULL
Traceback (most recent call last):
  File "/var/tmp/openupgrade/8.0/server/openerp/models.py", line 2702, in _auto_init
    cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, k))
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 158, in wrapper
    return f(self, *args, **kwargs)
  File "/var/tmp/openupgrade/8.0/server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "view_location_id" contains null values
```

hbrunn commented:
The workflow is as you describe, so the scripts should take care of fixing those values. If this is possible at all, there are some concepts that are simply new and there's no way to have a sensible default value.

You should btw always run the migration on a fresh database. Many of the migration scripts will break when being run multiple times

hbrunn commented on addons/stock/migrations/8.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,598 @@
+    stock_move_obj.action_done(cr, uid, done_move_ids, context=None)
+
+
+def delete_stock_account_journal(cr):
```
This should be an xmlid rename instead of deleting the journal. We'll soon provide a fix for that

gfcapalbo committed:
[FIX] renamed xml id instead of deleting journal

Commit:
[FIX] Don't map track_production to track_all. Handle track_production in mrp
[FIX] Defer mapping procure_method of procurement orders until additional rules
have been created in mrp and purchase
[FIX] Defer mapping of product valuation until the field actually exists, and
take the default property value into account.

StefanRijnhart commented:
I've pushed some improvements and set back to work in progress to indicate that it needs more work IMHO.

StefanRijnhart commented on addons/stock/migrations/8.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,526 @@
+
+
+def _migrate_stock_warehouse(cr, id):
+    """Warehouse adaptation to the new functionality. Sequences, Picking types, Rules, Paths..
```
Note to self: check if this wad of copied code can simply be triggered by calling one or more methods of the warehouse model. Or better, reuse existing sequences and optionally keep sharing them between warehouses if that was the case in 7.0. NB sequence and sequence type codes 'stock.picking' (for internal pickings) were renamed in OCB to 'stock.picking.internal'. Match both when retrieving existing sequences.

Commit:
[UPD] work text

Commit:
[FIX] Don't migrate the wrong field to the procurement's reservation (twice!)
[RFR] Don't converge obsolete procurement state in this module (and wrongly).
Has been done in the procurement migration by now.
[RFR] Centralize check on procurement migration or fresh installation

Commit:
[RFT] Rewrite warehouse migration based on known defaults
[IMP] Preserve counter of existing sequences
[IMP] Extend rename_tables API with magic argument None
[IMP] Add color, multiwarehouse sequence logic
[RFT] Rewrite move/inventory link migration in SQL and actually call it
[RFT] Streamline uid/registry arguments

Commit:
[IMP] Fixes and improved logging

Commit:
[IMP] Allow loading yml in openupgrade.load_data

Commit:
[FIX] Re-establish dynamically created XML IDs removed in previous commit, this time by loading the original yml file after updating warehouse configuration

Commit:
[FIX] Reloading yaml does not update the res_ids

Commit:
[IMP] Finalize stock inventory migration

Commit:
[DEL] Default location on inventories already set in pre script
[ADD] Warn about warehouses picked for location paths

Commit:
[FIX] location_id can be filled if type was changed afterwards

Commit:
[IMP] Sync stock move warehouse with the one from its procurement

Commit:
[RFR] Assigning warehouses and locations

Commit:
[IMP] Docstrings

hbrunn committed:
[FIX] typo

hbrunn committed:
[FIX] don't crash if no rule is found

legalsylvain commented:
Hi, I'd like to work on 'stock' migration. @StefanRijnhart, you put a WIP label, but you commited a lot after. 
Is this PR ready to review ? Is there some work to do ?
Thanks.

StefanRijnhart commented:
So the status AFAIK is: seems complete for some scenarios, but may be dubious for other, untested scenarios. Think multiwarehouse, multicompany, nontrivial chained picking steps. Other than that, I have no concrete todos in mind except that due to the complexity of this migration, a full revalidation of the work would be useful.

legalsylvain commented:
All right, thanks a lot.
I'll work on that in the next weeks, depending of a pending lead.
Regards.

legalsylvain commented:
All right, thanks a lot.
I'll work on that in the next weeks, depending of a pending lead.
Regards.

legalsylvain committed:
Merge branch '8.0' of https://github.com/OpenUpgrade/OpenUpgrade into 8.0-stock-onesteinbv

legalsylvain committed:
[flake8]

legalsylvain committed:
[FIX] flake8;

legalsylvain committed:
[FIX] flake8;

thomaspaulb commented:
I got the following error while running this script:

```
File "/var/tmp/openupgrade/8.0/addons/stock/migrations/8.0.1.1/post-migration.py", line 561, in migrate_product_supply_method
procure_method_legacy, ('make_to_order',)))
TypeError: not all arguments converted during string formatting
```

To solve it I needed to make the following change in post_migration.py:

```
cr.execute("SELECT id FROM product_template WHERE %s = %%s" % (
procure_method_legacy, ('make_to_order',)))
```

To

```
cr.execute("SELECT id FROM product_template WHERE %s = '%s'" % (
procure_method_legacy, 'make_to_order')
```

hbrunn commented:
please provide a PR against this branch

Commit:
Resolves "TypeError: not all arguments converted during string formatting" on line 561 of post_migration.py

Commit:
flake8

thomaspaulb commented:
https://github.com/OpenUpgrade-dev/OpenUpgrade/pull/2

Commit:
Still have to use psycopg safe parameter substitution

pedrobaeza committed:
Merge pull request #2 from antiflu/8.0-stock-antiflu

Resolves "TypeError: not all arguments converted during string formattin...

njeudy commented on addons/stock/migrations/8.0.1.1/pre-migration.py:
```diff
@@ -0,0 +1,164 @@
+        'stock_account.property_stock_account_output_prd'),
+    ('stock.property_stock_journal', 'stock_account.property_stock_journal'),
+    ('stock.stock_journal', 'stock_account.stock_journal'),
+]
```
Think xmlid moved to stock_account module should be rename in pre-migration script of stock_account and not here.
stock_account is not a depends of stock ..

legalsylvain replied:
I'm not sure.
Upgrade of 'stock' will delete xml_ids that disappeared. (except noupdate xml_id)
When process will update 'stock_account', the object has been deleted.

Regards.

legalsylvain commented on addons/stock/migrations/8.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,714 @@
+        """)
+    company_obj = registry['res.company']
+    warehouse_obj = registry['stock.warehouse']
+    procurement_obj = registry['stock.warehouse']
```
:question: I guess it should be procurement_obj = registry['procurement.order'] ?
My current log (#l685), indicate :
stock.warehouse.write() with unknown fields: warehouse_id

StefanRijnhart replied:
Yes, I'm afraid so ;-)

pedrobaeza replied:
Fix this part

pedrobaeza commented:
I have face one error with this:

```
2015-03-23 03:25:04,195 29802 ERROR nayar_migrated OpenUpgrade: stock: error in migration script /opt/odoo/v8/OpenUpgrade/addons/stock/migrations/8.0.1.1/post-migration.py: maximum recursion depth exceeded while calling a Python object
2015-03-23 03:25:04,249 29802 ERROR nayar_migrated OpenUpgrade: maximum recursion depth exceeded while calling a Python object
Traceback (most recent call last):
  File "/opt/odoo/v8/OpenUpgrade/openerp/openupgrade/openupgrade.py", line 807, in wrapped_function
    func(cr, version)
  File "/opt/odoo/v8/OpenUpgrade/addons/stock/migrations/8.0.1.1/post-migration.py", line 703, in migrate
    migrate_stock_qty(cr, registry)
  File "/opt/odoo/v8/OpenUpgrade/addons/stock/migrations/8.0.1.1/post-migration.py", line 616, in migrate_stock_qty
    stock_move_obj.action_done(cr, uid, done_move_ids, context=None)
  File "/opt/odoo/v8/OpenUpgrade/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/opt/odoo/v8/OpenUpgrade/addons/stock/stock.py", line 2402, in action_done
    self.write(cr, uid, ids, {'state': 'done', 'date': time.strftime(DEFAULT_SERVER_DATETIME_FORMAT)}, context=context)
  File "/opt/odoo/v8/OpenUpgrade/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/opt/odoo/v8/OpenUpgrade/addons/stock/stock.py", line 1968, in write
    self.write(cr, uid, [move.move_dest_id.id], propagated_changes_dict, context=context)
  File "/opt/odoo/v8/OpenUpgrade/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
```

I'm checking

pedrobaeza committed:
[FIX] stock: model name

legalsylvain committed:
[FIX] Incorrect migration of stock_warehouse code field; Using orm function for update the field to let the end of script to access to the value ;

legalsylvain committed:
[REF] 'stock' migration. Add a logger call for quants process;

legalsylvain commented:
Hi, 
I'm facing a problem, working on stock migration on a middle sized database. The DB has ~170.000 stock moves in a 'done' state. 
The call of the function 'migrate_stock_qty' to generate quants take a long time : > 10 hours, and it didn't finished. 2/3 of stock moves managed (I added some debug log in action_done function).
1/ @StefanRijnhart, @pedrobaeza, @other Did you faced the same problem ? Maybe there is some bug in my DB but I don't think so.
2/ Do you think it could be possible to realize a SQL request to create quants instead of the current trick (set all stock moves in draft state and call action_done).
3/ other ORM ideas to reduce the duration ? 

Thanks for your answer.

pedrobaeza commented:
@legalsylvain, I also faced a problem that I have put before. I'm working on a SQL in OpenUpgrade API method, but frozen for now until next week.

legalsylvain commented:
Yes, I saw it, but in my case I don't have recursion problem. How many stock move do you have and how long does this process (before your error) ?
Other point : in action_done function, the last call is : 

```
self.write(cr, uid, ids, {'state': 'done', 'date': time.strftime(DEFAULT_SERVER_DATETIME_FORMAT)}, context=context)
```

I don't think that using current server date is a good point. 
Maybe we could work together on SQL alternative ?

pedrobaeza commented:
I was in a similar number of records, and it took about 20 h, so it's good to have an SQL alternative. One of the bottleneck is also m2o_to_x2m function, but there's a lot of work to do.

Probably I can't work this week on this, but if you want to start...

legalsylvain commented:
Hi @all,
I was talking with @pedrobaeza. I know that what I'll propose you is not a very regular process but I have a problem with this PR. It is opened since October, 28 2014.
I think that it is too big to release. In other hand, it globally works. So, to avoid to block openupgrade process, I propose you to realize the Merge of this PR, and after to realize some little PR, more easy to review to improve duration (my case), or to fix some very specific case unmanaged for the time being. 
We could mark stock as 'done (experimental)' in documentation until more reviewers consider it as mature. 
How do you think about that ?

Regards.

paulius-sladkevicius commented:
I agree with @legalsylvain, this is open too long and looks already used to migrate db's, experimental sounds good. Could be it will make more attraction to make migrations and push fixes based on this great work then trying to work on this PR.

legalsylvain commented:
@hbrunn , @StefanRijnhart , @pedrobaeza  @all are you against this point of view.
If not, I'll merge this PR at the end of the week, setting this module as 'done (experimental)' and work on a little PR to improve quants creation. (reduce script duration)

thomaspaulb commented:
I already used it and for me it worked, there are probably still some glitches somewhere but they can be fixed whenever they pop up I think?

StefanRijnhart commented:
Yes, merging with experimental or alpha status is fine.

pedrobaeza commented:
I merge then and put it as experimental
